### PR TITLE
Unidirectional wgbs read reordering

### PR DESCRIPTION
**Pull request by vinjana**
Analysis of undirectional WGBS data can improved by estimating the most likely orientation. To this aim the already existing `WGBS_read_pair_reconstruction.py` script was updated and integrated into the workflow

+ extensive refactoring including a solution for deconvoluting pipe-chains
+ import of code from the BamToFastqWorkflow (on the long run this library code should maybe go to another library or the DefaultPlugin -- but not now ... no time)

### CODE REVIEW

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh [hunk 1]
```diff
@@ -42,80 +46,87 @@ tempFlagstatsFile=${FILENAME_FLAGSTATS}.tmp
...
+    fi
+
+    if [[ "${reorderUndirectionalWGBSReadPairs:-false}" == "true" ]]; then
+        extendPipePair "$fqName" "reorder" -- reorderUndirectionalReads
+    fi
+
+    extendPipe $(mkPairedPipeName 1 "$fqName") "fqconv" -- methylCfqconv 1
+	extendPipe $(mkPairedPipeName 2 "$fqName") "fqconv" -- methylCfqconv 2
+
+    INPUT_PIPES=""
```
Comment (vinjana) [inline resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: The next code can probably be further DRYed :-P

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh [hunk 2]
```diff
@@ -42,80 +46,87 @@ tempFlagstatsFile=${FILENAME_FLAGSTATS}.tmp
...
+true & procUnpack2=$!
+
+fqName="fastq"
+mkPipePairSource "$fqName"
+if [[ "$bamFileExists" == "false" ]]; then	# we have to make the BAM
+    getFastqAsciiStream "$RAW_SEQ_1" > $(getPairedPipeEndPath 1 "$fqName") & procUnpack1=$!
+    getFastqAsciiStream "$RAW_SEQ_2" > $(getPairedPipeEndPath 2 "$fqName") & procUnpack2=$!
+
+    if [[ "$qualityScore" == "illumina" ]]; then
+	    extendPipe $(mkPairedPipeName 1 "$fqName") "qScore" -- toIlluminaScore
```
Comment (dankwart-de) [inline resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: indentation?

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh [hunk 3]
```diff
@@ -42,80 +46,87 @@ tempFlagstatsFile=${FILENAME_FLAGSTATS}.tmp
...
+    if [[ "$useAdaptorTrimming:-false" == "true" ]]; then
+        extendPipePair "$fqName" "trimmomatic" -- trimmomatic
+    fi
+
+    if [[ "${reorderUndirectionalWGBSReadPairs:-false}" == "true" ]]; then
+        extendPipePair "$fqName" "reorder" -- reorderUndirectionalReads
+    fi
+
+    extendPipe $(mkPairedPipeName 1 "$fqName") "fqconv" -- methylCfqconv 1
+	extendPipe $(mkPairedPipeName 2 "$fqName") "fqconv" -- methylCfqconv 2
```
Comment (dankwart-de) [inline resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: indentation again.

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh [hunk 4]
```diff
@@ -42,80 +46,87 @@ tempFlagstatsFile=${FILENAME_FLAGSTATS}.tmp
...
+
+    INPUT_PIPES=""
+    if [[ ${LENGTH_SEQ_1} == 0 ]]; then
+        cat "$(getPairedPipeEndPath 1 $fqName)" > /dev/null
+    else
+	    INPUT_PIPES="$(getPairedPipeEndPath 1 $fqName)"
```
Comment (dankwart-de) [inline resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: indentation

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh [hunk 5]
```diff
@@ -42,80 +46,87 @@ tempFlagstatsFile=${FILENAME_FLAGSTATS}.tmp
...
+    fi
+
+    if [[ ${LENGTH_SEQ_2} == 0 ]]; then
+        cat "$(getPairedPipeEndPath 2 $fqName)" > /dev/null
+    else
+	    INPUT_PIPES="${INPUT_PIPES} $(getPairedPipeEndPath 2 $fqName)"
```
Comment (dankwart-de) [inline resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: indentation

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh (hunk 1 above)
Comment (dankwart-de) [reply resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: Maybe put in a method and call this with 1 and 2.

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh [hunk 6]
```diff
@@ -42,80 +46,87 @@ tempFlagstatsFile=${FILENAME_FLAGSTATS}.tmp
...
+    fi
```
Comment (dankwart-de) [inline resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: Really nice to read!

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 7]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
     echo "$result"
 }
 
+normalizeBoolean() {
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Issue to put this in a new DefaultPlugin version?

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 8]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
...
+
+listPipeEnds() {
+    for i in "${!_pipeEnds[@]}"; do
+        echo "'$i' => '${_pipeEnds[$i]}'"
+    done
+}
+
+# Make a path without registering it.
+mkPipePath() {
+    local pipeName="${1:?Named-pipe base name}"
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Maybe a better error message? "Named pipe cannot be created: Call mkPipePath("somePipesName")" or so?

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 9]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
...
+    echo "$_pipePath/$pipeName"
+}
+
+# Just get the path from the registry.
+getPipeEndPath() {
+    local pipeName="${1:?No pipename}"
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Also here: "Call getPipeEndPath("somePipeName")"?

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 10]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
...
+    if [[ ! ${_pipeEnds[$pipeName]+_} ]]; then
+        throw 152 "Could not find pipe by name '$pipeName'"
+    fi
+    local pipePath="${_pipeEnds[$pipeName]}"
+    echo "$pipePath"
+}
+
+# Register a path in the registry under the key "pipeName".
+setPipeEndPath() {
+    local pipeName="${1:?No pipename}"
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Again, more helpful error messages.

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 11]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
...
+    _pipeEnds[$pipeName]="$pipePath"
+}
+
+# Create a path from the name and the tag and register it as new pipe-end in the registry.
+# Note that the pipe is registered as tempfile and will be deleted upon cleanUp.
+updatePipeEndPath() {
+    local pipeName="${1:?Named-pipe base name}"
+    local tag="${2:?Pipe tag}"
+    local pipePath=$(mkPipePath "${pipeName}_$tag")
+    mkfifo "$pipePath" || throw 150 "Could not create named pipe at '$pipePath'"
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Will the old pipe path be deleted?

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 12]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
...
+#####################################################################
+# The pipe-extension API follows this pattern:
+#
+# * pipeExtenderFunction pipeBaseName pipeStepExtension -- command
+#
+# * The pipeBaseName is the name under which the pipe (or derived from this: pipe pair) is registered.
+# * The pipeStepExtension is used to tag the actual pipe used in the specific step. Use a name describing the content flowing through the pipe,
+#   i.e. the output of the extension step you are declaring.
+# * Current pipe extender functions are extendPipe (for single pipes) and extendPipePair (for r1/r2 pipes)
+# * To work with paired pipes (r1/r2) mkPairedPipeName is used to calculate the r1_ or r2_ pipeBaseName
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Maybe put in:

- Note, that we use associative arrays. To understand the code below, please read: https://blabla

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 13]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
...
+#
+# The "--" is optional.
+extendPipe() {
+    local pipeName="${1:?No pipe basename}"
+    local tag="${2:?No tag}"
+    shift 2
+    if [[ "$1" == "--" ]]; then
+        shift
+    fi
+    local command="${1:?No command/function}"
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: $1 again?

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 14]
```diff
@@ -124,6 +124,216 @@ stringJoin () {
...
+    shift
+    declare -a args=("$@")
+
+    local inpipe=$(getPipeEndPath "$pipeName")
+    updatePipeEndPath "$pipeName" "$tag"
+    local outpipe=$(getPipeEndPath "$pipeName")
+
+    "$command" "$inpipe" "$outpipe" "${args[@]}" & registerPid
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Does refer command to a binary or to a function? I don't see, how this works for all binaries.

File: resources/tests/analysisTools/qcPipeline/bashLib.sh [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+teardownPipePath() {
+    rm -rf "$_pipePath"
+    unset _pipePath
+    unset _pipeEnds
+    cleanUp_BashSucksVersion
+}
+
+testSetGetPipeEndPath() {
+    setupPipePath
+    assertFalse "test getPipeEndPath 'doesnotexist'"
```
Comment (dankwart-de) [inline resources/tests/analysisTools/qcPipeline/bashLib.sh]: Is that right?
assertFalse "test" "$(getPipeEndPath 'doesnotexist'"?

File: resources/tests/analysisTools/qcPipeline/chrom-sizes-file.tsv [hunk 16]
```diff
@@ -0,0 +1,5 @@
+1	1
```
Comment (dankwart-de) [inline resources/tests/analysisTools/qcPipeline/chrom-sizes-file.tsv]: 5 Lines? That right?

File: README.md [hunk 17]
```diff
@@ -1,8 +1,42 @@
...
-- PanCancer alignment workflow for whole genome and exome
-- Postmerge QC workflow for whole genome and exome
-- Bisulfite core workflow
+This plugins contains alignment and quality control related [Roddy](https://github.com/TheRoddyWMS/Roddy) workflows.
 
+This is the branch for 1.2.73 maintenance branch with only minimal documentation. Please see the master branch for details.
 
+## Read-Reordering with Unidirectional WGBS Data
+
+By setting `reorderUnidirectionalWGBSReadPairs` the read-reordering script will be run that decides based on the relative frequencies of TC and AG dinucleotides in both reads, what is the most likely correct orientations of the reads, and may swap the two reads.
```
[the remark this answers is not in the document]
Comment (vinjana) [reply README.md]: The tool determines the order.

I now point out that the explanation can be found with the original script.

File: README.md [hunk 18]
```diff
@@ -1,8 +1,42 @@
...
+
+Note that after the swapping, the read-numbers are reversed. What was R1 in the input FASTQ will be R2 in the output BAM, and vice versa.
+
+Furthermore, not all reads can be unambiguously classified. These unclassified reads are currently dropped.
```
[the remark this answers is not in the document]
Comment (vinjana) [reply README.md]: Not yet. See #47 .

File: README.md [hunk 19]
```diff
@@ -1,8 +1,42 @@
...
 
+## Read-Reordering with Unidirectional WGBS Data
+
+By setting `reorderUnidirectionalWGBSReadPairs` the read-reordering script will be run that decides based on the relative frequencies of TC and AG dinucleotides in both reads, what is the most likely correct orientations of the reads, and may swap the two reads.
+
+Note that after the swapping, the read-numbers are reversed. What was R1 in the input FASTQ will be R2 in the output BAM, and vice versa.
+
+Furthermore, not all reads can be unambiguously classified. These unclassified reads are currently dropped. 
+
+The original script can be found [here](https://github.com/cimbusch/TWGBS.git).
```
[the remark this answers is not in the document]
Comment (vinjana) [reply README.md]: Done.

File: resources/analysisTools/bisulfiteWorkflow/WGBS_read_pair_reconstruction.py [hunk 20]
```diff
@@ -1,4 +1,9 @@
-#!/usr/bin/python
+#!/usr/bin/env python
+#
+# Copyright (c) 2018 German Cancer Research Center (DKFZ).
+#
+# Distributed under the MIT License (license terms are at https://github.com/cimbusch/TWGBS).
+#
```
[the remark this answers is not in the document]
Comment (vinjana) [reply resources/analysisTools/bisulfiteWorkflow/WGBS_read_pair_reconstruction.py]: Done.

File: resources/analysisTools/qcPipeline/bashLib.sh (hunk 12 above)
Comment (vinjana) [reply resources/analysisTools/qcPipeline/bashLib.sh]: Done.

File: resources/analysisTools/qcPipeline/bashLib.sh (hunk 13 above)
Comment (vinjana) [reply resources/analysisTools/qcPipeline/bashLib.sh]: Yes. I did shifts before!

File: resources/analysisTools/qcPipeline/bashLib.sh (hunk 14 above)
Comment (vinjana) [reply resources/analysisTools/qcPipeline/bashLib.sh]: Currently, a binary or bash function.

Maybe it is possible to say
```bash
$command "$inpipe" ...
```
to allow for a "binary -arg1 -arg2" argument as command that is then interpreted as command prefix. I don't know and could try later.

File: resources/tests/analysisTools/qcPipeline/chrom-sizes-file.tsv (hunk 16 above)
Comment (vinjana) [reply resources/tests/analysisTools/qcPipeline/chrom-sizes-file.tsv]: This file was for the grouped qc summary that is from 1.3.0. I forgot to delete the file after backporting the whole test code from there.

File: resources/analysisTools/qcPipeline/bashLib.sh (hunk 14 above)
Comment (vinjana) [reply resources/analysisTools/qcPipeline/bashLib.sh]: Even though this is rare, it is possible that the path to command contains a space. Without the quotes that would not resolve correctly. I rather request from the user to have the desired interface "command inpipe outpipe", even if he will have to define a function to achieve this.

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh (hunk 2 above)
Comment (vinjana) [reply resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: Was not visible in Idea, but really broken. Fixed!

File: resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh [hunk 21]
```diff
@@ -34,59 +35,61 @@ MBUF_LARGE="${MBUFFER_BINARY} -m ${MBUFFER_SIZE_LARGE} -q -l /dev/null"
...
+if [[ $LENGTH_SEQ_1 -eq 0 ]] && [[ $LENGTH_SEQ_2 -eq 0 ]]; then
     throw 1 "Both input files are empty: '$RAW_SEQ_1' and '$RAW_SEQ_2'"
+elif [[ $LENGTH_SEQ_1 -eq 0 ]] || [[ $LENGTH_SEQ_2 -eq 0 ]]; then
+    useSingleEndProcessing=true
 fi
 
+
 # Determine the quality score
-if [[ $LENGTH_SEQ_1 != 0 ]]; then
+set +e   # The following few lines fail with exit 141 due to unknown reasons is `set -e` is set.
```
Comment (dankwart-de) [inline resources/analysisTools/bisulfiteWorkflow/bwaMemSortSlimWithReadConversionForBisulfiteData.sh]: is ... is set => if ... is set

File: resources/analysisTools/qcPipeline/bashLib.sh [hunk 22]
```diff
@@ -178,6 +184,7 @@ waitForRegisteredPids_BashSucksVersion() {
     jobs
     declare -a realPids=$(listPids)
     if [[ -v realPids && ${#realPids[@]} -gt 0 ]]; then
+        # TODO Make this a look and report the exact pid that failed (or the key, after switching from array to dictionary).
```
Comment (dankwart-de) [inline resources/analysisTools/qcPipeline/bashLib.sh]: Make this a look?

**Pull request merged**